backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/util/GitUtils.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SonarLint Core - Commons
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.sonarlint.core.commons.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.eclipse.jgit.ignore.FastIgnoreRule;
+import org.eclipse.jgit.ignore.IgnoreNode;
+import org.eclipse.jgit.lib.Repository;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+
+import static org.eclipse.jgit.lib.Constants.GITIGNORE_FILENAME;
+import static org.sonarsource.sonarlint.core.commons.util.gitblame.GitBlameUtils.buildGitRepository;
+
+public class GitUtils {
+
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+
+  private GitUtils() {
+    // Utility class
+  }
+
+  /**
+   * Assumes the supplied {@param baseDir} or some of its parents is a git repository.
+   * The method checks {@param filePathsToAnalyze} against patterns defined in the .gitignore file and returns
+   * those that are not ignored.
+   * If error occurs during execution then all {@param filePathsToAnalyze} are considered as not ignored.
+   */
+  public static List<URI> filterOutIgnoredFiles(Path baseDir, List<URI> filePathsToAnalyze) {
+    try {
+      var gitRepo = buildGitRepository(baseDir);
+      var gitRepoRelativeProjectBaseDir = gitRepo.getWorkTree().toPath().relativize(baseDir);
+      var ignoreNode = buildIgnoreNode(gitRepo);
+      return ignoreNode.getRules().isEmpty() ? filePathsToAnalyze : filePathsToAnalyze.stream().filter(uri -> {
+        var fileRelativeToGitRepoPath = gitRepoRelativeProjectBaseDir.resolve(Path.of(uri)).toUri().toString();
```

Comment:
I think we could remove `.toUri()` part just before converting toString. I am not sure it is necessary for Windows, though.

Reply:
Well spotted....yep, it's for Windows as it works fine with URI which is the same as for unix (with the forward slash as opposed to Path)